src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    let nextState = { ...currentState };
+
+    if (action.type === 'addProperties') {
+      nextState = { ...nextState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      action.keysToRemove.forEach((key) => {
+        delete nextState[key];
+      });
+    }
+
+    if (action.type === 'clear') {
+      nextState = {};
+    }
+
+    stateHistory.push(nextState);
```

Comment:
[DONT REPEAT YOURSELF] - You are pushing the `nextState` object to the `stateHistory` array in each iteration of the loop. This operation is the same for all cases, so you can move it outside of the `if` or `switch` block, just before the end of the `forEach` callback function.